internals/cli/cmd_run.go

```diff
@@ -202,13 +211,24 @@ func runDaemon(rcmd *cmdRun, ch chan os.Signal, ready chan<- func()) error {
 		} else {
 			logger.Noticef("Started default services with change %s.", changeID)
 		}
-	}
 
-	var stop chan struct{}
-	if ready != nil {
-		stop = make(chan struct{}, 1)
-		ready <- func() { close(stop) }
-		close(ready)
```

Comment:
Yeah, I'm having a bit of trouble following the logic too. Seems weird to be passing on a channel a function that closes another channel.

Somewhat related: https://github.com/canonical/pebble/issues/167 -- I wonder if `pebble run` should also wait for services to be started successfully before starting up (or if not waiting, exit if they don't start properly). It seems like it'd be better for that to be a hard error. But maybe that's a separate PR/discussion...

Reply:
This was introduced in #188 while adding the `enter` command. The commits added an argument to the `rcmd.run` function in [`cmd_run.go`](https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_run.go#L82):
https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_run.go#L82
The `ready` channel is passed from [`cmd_enter.go`](https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_enter.go#L150-L157) and in other cases, it's typically `nil`. 
https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_enter.go#L150-L157
After running the `run()` function with the channel, execution [waited on that channel](https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_enter.go#L159-L166) and one other; the latter received values if there was any `panic()` while executing `run()`.
https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_enter.go#L159-L166
The `runReadyCh` in `cmd_enter.go` (`ready` in `cmd_run.go`) received a function to close the `stop` channel in `cmd_run.go`, after the daemon had started, the watchdog was run and the default services had begun in [`cmd_run.go`](https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_run.go#L207-L212).
https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_run.go#L207-L212
The `stop` channel, if it was closed or received any values would kill the daemon and stop the pebble server, as specified here:
https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_run.go#L229-L230
This behavior was utilized in [`cmd_enter.go`](https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_enter.go#L174-L176), to shut down the pebble daemon if there was no need to keep the "Service Manager" alive.
https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_enter.go#L174-L176

---

**New changes**

The desire is to send to the `ready` channel `cmd_run.go` after all the default services have "started". Previously, it only made an API call and signalled that the services have started, through the `ready` channel. This produced a few errors as I specified in https://github.com/canonical/pebble/pull/257#issue-1793475495. The new changes are supposed to avoid those scenarios and send to `ready` only after all the default services have started. This is being done by sending to the channel when the respective autostart "change" is "Ready".

In the new changes, after making the autostart POST call https://github.com/canonical/pebble/blob/8e96c74a21cc104ac722bf4160d7db6e11ddeb60/internals/cli/cmd_run.go#L199
a goroutine starts to wait on the changes to be "Ready".
```go
        if ready != nil {
            // wait for the default services to start
            autoStartReady = make(chan struct{}, 1)
            go func() {
                waitCmd := waitMixin{
                    hideProgress: true,
                }
                waitCmd.setClient(rcmd.client)
                _, err := waitCmd.wait(changeID)
                if err != nil {
                    logger.Debugf("Error starting default services: %v", err)
                }
                autoStartReady <- struct{}{}
            }()
        }
```
Only when the change has been "Ready", it sends to a new channel `autoStartReady`. Afterwards, the previous logic is used to send to the `ready` (`runReadyCh` in `cmd_enter.go`) channel, as before.
```go
        case <-autoStartReady:
            notifyReady()
        case <-stop:
            break out
```

Reply:
Thanks. Still seems a little convoluted, but an obviously-simpler approach is not jumping out at me, so seems reasonable.

What do you think of my earlier question/comment:

> Somewhat related: #167 -- I wonder if pebble run should also wait for services to be started successfully before starting up (or if not waiting, exit if they don't start properly). It seems like it'd be better for that to be a hard error. But maybe that's a separate PR/discussion...

Reply:
> Somewhat related: https://github.com/canonical/pebble/issues/167 -- I wonder if pebble run should also wait for services to be started successfully before starting up (or if not waiting, exit if they don't start properly). It seems like it'd be better for that to be a hard error. But maybe that's a separate PR/discussion...

Concerning `pebble run`, I think it's better to exit and report if the services don't start properly. In `enter`, we need the wait because the subcommand which follows needs to be run after the services have started. That may not be the case for `run` though. 

Also, I am not sure and please correct me if I am wrong, but it seemed to be that if a startup-enabled service failed to start with an error, the services that followed in the queue were not started as well. So I think it's best to report any errors while starting up and exit afterwards.

Reply:
I ran into a similar situation to #167 while adding the recent tests; I had to debug for some time before realizing "Oh wait, this must be that...! :man_facepalming:".
I echo my earlier comments. I don't think `pebble run` or `pebble enter --run` (without any subcommand) should wait for the services to start. Although I do somewhat understand the motivation in that. But I think, it's crucial that we report (and exit) if any default service fails to start.
In any case, it should be trivial to do so on top of this PR as a separate commit or even a new PR afterwards. Let me know what you think.